scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -794,52 +816,193 @@ static void update_sys_cpu_load(void)
+		stat_next->inc1k_low = ((LAVD_BOOST_RANGE >> 1) * 1000) /
+					(stat_next->avg_lat_cri -
+					 stat_next->min_lat_cri);
 	}
 
-	if ((cutil_next->max_lat_cri + 1) == cutil_next->avg_lat_cri)
-		cutil_next->inc1k_high = 0;
+	if ((stat_next->max_lat_cri + 1) == stat_next->avg_lat_cri)
+		stat_next->inc1k_high = 0;
 	else {	
-		cutil_next->inc1k_high = ((LAVD_BOOST_RANGE >> 1) * 1000) /
-					 (cutil_next->max_lat_cri + 1 -
-					  cutil_next->avg_lat_cri);
+		stat_next->inc1k_high = ((LAVD_BOOST_RANGE >> 1) * 1000) /
+					 (stat_next->max_lat_cri + 1 -
+					  stat_next->avg_lat_cri);
 	}
 
 	/*
 	 * Make the next version atomically visible.
 	 */
-	cutil_next->last_update_clk = now;
-	flip_sys_cpu_util();
+	stat_next->last_update_clk = now;
+	flip_sys_stat();
 }
 
-static int update_timer_fn(void *map, int *key, struct bpf_timer *timer)
+static u64 calc_nr_active_cpus(struct sys_stat *stat_cur)
+{
+	u64 nr_active;
+
+	/*
+	 * nr_active = ceil(nr_cpus_onln * cpu_util * per_core_max_util)
+	 */
+	nr_active  = (nr_cpus_onln * stat_cur->util * 1000) + 500;
+	nr_active /= (LAVE_TC_PER_CORE_MAX_CTUIL * 1000);
+
+	/*
+	 * If a few CPUs are particularly busy, boost the overflow CPUs by 2x.
+	 */
+	nr_active += min(LAVD_TC_NR_OVRFLW, (stat_cur->nr_violation) / 1000);
+	nr_active = max(min(nr_active, nr_cpus_onln),
+			LAVD_TC_NR_ACTIVE_MIN);
+
+	return nr_active;
+}
+
+static const struct cpumask *cast_mask(struct bpf_cpumask *mask)
+{
+	return (const struct cpumask *)mask;
+}
+
+static bool test_and_clear_cpu_periodically(u32 cpu,
+					    struct bpf_cpumask *cpumask)
+{
+	u32 clear;
+	bool ret = false;
+
+	/*
+	 * If the CPU is on, we clear the bit once every eight times
+	 * (LAVD_TC_CPU_PIN_INTERVAL_DIV). Hence, the bit will be
+	 * probabilistically cleared once every 200 msec (8 * 25 msec).
+	 */
+	if (!bpf_cpumask_test_cpu(cpu, cast_mask(cpumask)))
+		return false;
+
+	clear = !(bpf_get_prandom_u32() % LAVD_TC_CPU_PIN_INTERVAL_DIV);
+	if (clear)
+		ret = bpf_cpumask_test_and_clear_cpu(cpu, cpumask);
+	return ret;
+}
+
+static void do_core_compaction(void)
+{
+	struct sys_stat *stat_cur = get_sys_stat_cur();
+	struct cpu_ctx *cpuc;
+	struct bpf_cpumask *active, *ovrflw;
+	int nr_cpus, nr_active, nr_active_old, cpu, i;
+
+	bpf_rcu_read_lock();
+
+	/*
+	 * Prepare cpumasks.
+	 */
+	active = active_cpumask;
+	ovrflw = ovrflw_cpumask;
+	if (!active || !ovrflw) {
+		scx_bpf_error("Failed to prepare cpumasks.");
+		goto unlock_out;
+	}
+
+	/*
+	 * Assign active and overflow cores
+	 */
+	nr_active_old = stat_cur->nr_active;
+	nr_active = calc_nr_active_cpus(stat_cur);
+	nr_cpus = nr_active + LAVD_TC_NR_OVRFLW;
+	bpf_for(i, 0, nr_cpus_onln) {
+		if (i >= LAVD_CPU_ID_MAX)
+			break;
+
+		/*
+		 * Skip offline cpu
+		 */
+		// cpu = get_cpu_id_of_order(i);
+		cpu = cpu_order[i];
+		cpuc = get_cpu_ctx_id(cpu);
+		if (!cpuc || !cpuc->is_online) {
+			bpf_cpumask_test_and_clear_cpu(cpu, active);
+			bpf_cpumask_test_and_clear_cpu(cpu, ovrflw);
```

Comment:
Is there a reason you're using the RMW version of these? If not then I think you can just use `bpf_cpumask_clear_cpu()` and `bpf_cpumask_set_cpu()`.

Reply:
active CPU mask is need to be atomically updated since it can be updated in two places concurrently -- do_core_compaction() and ops.dispatch(). In my checking the code, bpf_cpumask_clear_cpu() and bpf_cpumask_set_cpu() do not have LOCK prefix so they do not guarantee the atomicity. So I used the CAS version of set/clear_cpu(). Does it make sense?

For the overflow mask, I can change it to use non-atomic version since there is no concurrent writes.

Reply:
I changed the update of overflow cpumask to non-atomic version.